resharper/resharper-unity/src/Rider/UnityRefreshTracker.cs

```diff
@@ -74,21 +77,24 @@ public Task Refresh(RefreshType force)
 
             task.ContinueWith(_ =>
             {
+                var list = new List<string> {solFolder.FullPath};
                 mySolution.Locks.ExecuteOrQueueEx(lifetimeDef.Lifetime, "RefreshPaths", () =>
                 {
-                    try
-                    {
-                        var list = new List<string> {solFolder.FullPath};
-                        solution.GetFileSystemModel().RefreshPaths.Start(new RdRefreshRequest(list, true));
-                    }
-                    finally
+                    var res = solution.GetFileSystemModel().RefreshPaths.StartAsTask(new RdRefreshRequest(list, true));
+                    res.ContinueWith(__ =>
                     {
-                        CurrentTask = null;
-                        lifetimeDef.Terminate();
-                    }
+                        mySolution.Locks.ExecuteOrQueueEx(lifetimeDef.Lifetime, "RefreshPaths", () =>
```

Comment:
Why ExecuteOrQueueEx? You can use TaskHost.GuardedMainThreadScheduler to continue on the main thread.